### PR TITLE
Dedup Lineage and getTableToExtract call in BigQuerySourceBase

### PR DESCRIPTION
**Pull request by Abacn**
Internal tests found after #31805, BigQueryIO export read (with query) will make an additional BigQuery query job.

This is due to getTableToExtract call will try to create a temp dataset and initiate a query job. Arguably this Lineage metrics call is not necessary, right above in L156,

https://github.com/apache/beam/blob/eadb81fd01f281755f18a8a3095e3c2c8194e9fc/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQuerySourceBase.java#L156

extractFiles already had same Lineage metrics call:

https://github.com/apache/beam/blob/eadb81fd01f281755f18a8a3095e3c2c8194e9fc/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQuerySourceBase.java#L124

**Please** add a meaningful description for your change here

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

**Comment by Abacn** [comment]
R: @ahmedabu98

**Comment by Abacn** [comment]
confirmed after this change only one query job is initiated (Dataflow job id: 2024-07-23_15_59_31-4881852833110454297)

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQuerySourceBase.java [hunk 1]
```diff
@@ -157,9 +157,6 @@ public List<BoundedSource<T>> split(long desiredBundleSizeBytes, PipelineOptions
       LOG.info("Extract job produced {} files", res.extractedFiles.size());
       if (res.extractedFiles.size() > 0) {
         BigQueryOptions bqOptions = options.as(BigQueryOptions.class);
-        // emit this table ID as a lineage source
-        Lineage.getSources()
-            .add(BigQueryHelpers.dataCatalogName(getTableToExtract(bqOptions), bqOptions));
```
Comment (ahmedabu98) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQuerySourceBase.java:162]: We can keep Lineage support and instead store the output of `getTableToExtract` in a private variable.

We can store the value when we perform `extractFiles`: https://github.com/apache/beam/blob/3d69dc97c41000d59d43a303912ac12d5f58dce3/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQuerySourceBase.java#L112-L114

Then here, if stored table reference is not null we can add it to Lineage

**Review by ahmedabu98 — APPROVED** [review]
LGTM

**Comment by Abacn** [comment]
two spanner change stream test failed, unrelated to this change

```
testOrderedWithinKey (org.apache.beam.sdk.io.gcp.spanner.changestreams.it.SpannerChangeStreamOrderedWithinKeyIT) failed

testInvalidRecordReceived (org.apache.beam.sdk.io.gcp.spanner.changestreams.SpannerChangeStreamErrorTest) failed
```

**Pull request merged**